File: core/src/io/github/jerukan/game/gameunits/unitdata/unitactions/MoveAction.java
package io.github.jerukan.game.gameunits.unitdata.unitactions;

import io.github.jerukan.game.GameState;
import io.github.jerukan.game.gameunits.Unit;
import io.github.jerukan.game.gameunits.unitdata.BaseUnit;
import io.github.jerukan.util.Constants;
import io.github.jerukan.util.Position;

import java.util.ArrayList;

public class MoveAction extends UnitAction {

    public MoveAction(BaseUnit baseUnit) {
        super(baseUnit);
        name = "move";
        speedConsumption = 0;   //speed is modified based on distance traveled
        requiresTarget = true;
    }

    @Override
    public void execute(Unit self) {

    }

    @Override
    public void execute(Unit self, Position target) {
        if(target.existsInArray(availableTargets)) {
            if(GameState.instance.unitState.positionAvailable(target)) {
                int speedconsump = 0;
                //checks the corresponding speed the tile will consume
                for(int i = 0; i < self.getAvailableTargets().size(); i++) {
                    if(self.getAvailableTargets().get(i).equals(target)) {
                        speedconsump = targetSpeedConsumptions.get(i);
                    }
                }
                self.setCurrentSpeed(self.getCurrentSpeed() - speedconsump);
                self.setPosition(new Position(target));
                self.moveSprite(target);
            }
        }
    }

    @Override
    public void getTarget(Unit self, Position startpos, Position checkedpos, ArrayList<Position> moves, ArrayList<Integer> moveConsump, int aggregateConsump, int movesleft) {
        int speedconsump = GameState.instance.boardState.tileFromPosition(checkedpos).getSpeedConsump();

        if(!startpos.equals(checkedpos)) {
            if (movesleft < speedconsump) {
                return;
            }

            aggregateConsump += speedconsump;
            movesleft -= speedconsump;

            if (!checkedpos.existsInArray(moves)) {
                moves.add(checkedpos);
                moveConsump.add(aggregateConsump);
            }
        }

        if (checkedpos.getX() + 1 < Constants.BOARD_WIDTH) {
            Position newpos = new Position(checkedpos.getX() + 1, checkedpos.getY());
            getTarget(self, startpos, newpos, moves, moveConsump, aggregateConsump, movesleft);
        }
        if (checkedpos.getX() - 1 >= 0) {
            Position newpos = new Position(checkedpos.getX() - 1, checkedpos.getY());
            getTarget(self, startpos, newpos, moves, moveConsump, aggregateConsump, movesleft);
        }
        if (checkedpos.getY() + 1 < Constants.BOARD_HEIGHT) {
            Position newpos = new Position(checkedpos.getX(), checkedpos.getY() + 1);
            getTarget(self, startpos, newpos, moves, moveConsump, aggregateConsump, movesleft);
        }
        if (checkedpos.getY() - 1 >= 0) {
            Position newpos = new Position(checkedpos.getX(), checkedpos.getY() - 1);
            getTarget(self, startpos, newpos, moves, moveConsump, aggregateConsump, movesleft);
        }
    }

    @Override
    public void generateTargets(Unit self) {
        availableTargets.clear();
        targetSpeedConsumptions.clear();
        getTarget(self, self.getPosition(), new Position(self.getPosition()), availableTargets, targetSpeedConsumptions, 0, self.getCurrentSpeed());
    }
}

File: core/src/io/github/jerukan/game/gameunits/unitdata/unitactions/MoveAction.kt
package io.github.jerukan.game.gameunits.unitdata.unitactions

import io.github.jerukan.game.GameState
import io.github.jerukan.game.gameunits.Unit
import io.github.jerukan.game.gameunits.unitdata.BaseUnit
import io.github.jerukan.util.Constants
import io.github.jerukan.util.Position

import java.util.ArrayList

class MoveAction(baseUnit: BaseUnit) : UnitAction(baseUnit) {

    init {
        name = "move"
        speedConsumption = 0   //speed is modified based on distance traveled
        requiresTarget = true
    }

    override fun execute(self: Unit) {

    }

    override fun execute(self: Unit, target: Position) {
        if (target.existsInArray(availableTargets)) {
            if (GameState.instance.unitManager.positionAvailable(target)) {
                var speedconsump = 0
                //checks the corresponding speed the tile will consume
                for (i in 0 until self.getAvailableTargets().size) {
                    if (self.getAvailableTargets()[i].equals(target)) {
                        speedconsump = targetSpeedConsumptions[i]
                    }
                }
                self.currentSpeed = self.currentSpeed - speedconsump
                self.position = Position(target)
                self.moveSprite(target)
            }
        }
    }

    override fun getTarget(self: Unit, startpos: Position, checkedpos: Position, moves: ArrayList<Position>, moveConsump: ArrayList<Int>, aggregateConsump: Int, movesleft: Int) {
        var aggregateConsump = aggregateConsump
        var movesleft = movesleft
        val speedconsump = GameState.instance.boardManager.tileFromPosition(checkedpos).speedConsump

        if (!startpos.equals(checkedpos)) {
            if (movesleft < speedconsump - 1) {
                return
            }

            if (!checkedpos.existsInArray(moves)) {
                moves.add(checkedpos)
                moveConsump.add(aggregateConsump)
            }
        }
        aggregateConsump += speedconsump
        movesleft -= speedconsump

        if (checkedpos.x + 1 < Constants.BOARD_WIDTH) {
            val newpos = Position(checkedpos.x + 1, checkedpos.y)
            getTarget(self, startpos, newpos, moves, moveConsump, aggregateConsump, movesleft)
        }
        if (checkedpos.x - 1 >= 0) {
            val newpos = Position(checkedpos.x - 1, checkedpos.y)
            getTarget(self, startpos, newpos, moves, moveConsump, aggregateConsump, movesleft)
        }
        if (checkedpos.y + 1 < Constants.BOARD_HEIGHT) {
            val newpos = Position(checkedpos.x, checkedpos.y + 1)
            getTarget(self, startpos, newpos, moves, moveConsump, aggregateConsump, movesleft)
        }
        if (checkedpos.y - 1 >= 0) {
            val newpos = Position(checkedpos.x, checkedpos.y - 1)
            getTarget(self, startpos, newpos, moves, moveConsump, aggregateConsump, movesleft)
        }
    }

    override fun generateTargets(self: Unit) {
        availableTargets.clear()
        targetSpeedConsumptions.clear()
        getTarget(self, self.position, Position(self.position), availableTargets, targetSpeedConsumptions, 0, self.currentSpeed)
    }
}